### ISSUE TITLE
preload_indices is not thread safe

### ISSUE THREAD

**Issue by madmax** [pat/thinking-sphinx#943]
Thinking sphinx should load all indexes when boot application but it don't load distributed indexes. 

It means we have 1 instance of configuration shared across threads with preload_indices == false

If 2 threads want upgrade index (for example real time index) at the same time we might have race conditions with preload_indices method. 

At the end we can have distributed indexes loaded twice with mean we have difrend document_id_for_key and this generates duplications 

example: http://mad.xan.pl/s/9vjTiKjGq2/

- We should load all indexes before we start threads
- It will be nice to have some sort controll how document_id_for_key is generated because currently if we add new index we need reindex everything (for real time index it might be long time). If it will be possible to specify somewere in config number of supported indexes (we can then give bigger number there and have some space to add new indexes without reindexing everything).

**Comment by pat** [comment]
Just pushed a commit to deal with the thread-safety: 98a513741e1bd4d1c9ce89e23615f228c64855ba. To give it a spin:

```ruby
gem 'thinking-sphinx', '~> 3.1.4',
  :git    => 'git://github.com/pat/thinking-sphinx.git',
  :branch => 'develop',
  :ref    => '98a513741e'
```

As for deeper level control over document_id_for_key… I'll keep it in mind, though not sure what the most elegant way to take care of it is.
